File: dot.most.box/src/sse.mjs
// SSE rooms: roomId -> Set<{ id: string, send: (data:any)=>void, res: ServerResponse }>
const rooms = new Map();

// ---- SSE helpers (deduplicated) ----
const makeSSESender = (rawRes) => (obj) => {
  try {
    rawRes.write(`data: ${JSON.stringify(obj)}\n\n`);
  } catch (_) {
    // ignore write errors
  }
};
const broadcastToRoom = (roomId, message, { excludeId, toId } = {}) => {
  const set = rooms.get(roomId);
  if (!set || set.size === 0) return 0;
  let delivered = 0;
  set.forEach((client) => {
    if (excludeId && client.id === String(excludeId)) return;
    if (toId && client.id !== String(toId)) return;
    client.send(message);
    delivered++;
  });
  return delivered;
};

// ============ HTTP + SSE 信令 ============
export const registerSSE = (server) => {
  // 订阅房间消息 SSE
  server.get("/sse.signaling", (request, reply) => {
    try {
      const { roomId, clientId } = request.query || {};
      if (!roomId || !clientId) {
        reply.code(400);
        reply.send({ ok: false, message: "roomId 和 clientId 必填" });
        return;
      }

      // SSE 必要响应头
      reply.raw.setHeader("Content-Type", "text/event-stream");
      reply.raw.setHeader("Cache-Control", "no-cache");
      reply.raw.setHeader("Connection", "keep-alive");
      reply.raw.setHeader("Access-Control-Allow-Origin", "*");
      reply.raw.flushHeaders?.();

      // 写入工具
      const send = makeSSESender(reply.raw);

      // 加入房间
      let set = rooms.get(roomId);
      if (!set) {
        set = new Set();
        rooms.set(roomId, set);
      }

      const client = { id: String(clientId), send, res: reply.raw };
      set.add(client);
      // console.log(`[SSE] client ${clientId} joined room ${roomId}, size=${set.size}`);

      // 初次问候
      send({ type: "hello", roomId, clientId, ts: Date.now() });

      // 心跳保持
      const heartbeat = setInterval(() => {
        try {
          reply.raw.write(`: ping ${Date.now()}\n\n`);
        } catch {}
      }, 15000);

      // 断开清理
      request.raw.on("close", () => {
        clearInterval(heartbeat);
        const s = rooms.get(roomId);
        if (s) {
          s.delete(client);
          if (s.size === 0) rooms.delete(roomId);
        }
        // console.log(`[SSE] client ${clientId} left room ${roomId}`);
      });
    } catch (error) {
      // console.error("[SSE] subscribe error:", error);
      reply.code(500).send({ ok: false, message: error.message });
    }
  });

  // 获取房间用户列表 HTTP
  server.get("/api.room", async (request, reply) => {
    try {
      const { roomId } = request.query || {};

      if (!roomId) {
        reply.code(400);
        return { ok: false, message: "roomId 必填" };
      }

      const set = rooms.get(roomId);
      const users = set ? Array.from(set).map((client) => client.id) : [];

      return { ok: true, users };
    } catch (error) {
      // console.error("[SSE] get room users error:", error);
      reply.code(500);
      return { ok: false, message: error.message };
    }
  });

  // 发送信令 HTTP
  server.post("/api.signaling", async (request, reply) => {
    try {
      const { roomId, from, to, type, payload } = request.body || {};

      if (!roomId || !from || !type) {
        reply.code(400);
        return { ok: false, message: "roomId, from, type 必填" };
      }

      const message = { from, type, payload, ts: Date.now() };
      const delivered = broadcastToRoom(roomId, message, {
        excludeId: from,
        toId: to,
      });

      return { ok: true, delivered };
    } catch (error) {
      // console.error("[SSE] post signaling error:", error);
      reply.code(500);
      return { ok: false, message: error.message };
    }
  });
};

File: dot.most.box/internal/sse/sse.go
// Package sse 提供服务器发送事件（Server-Sent Events）信令通道。
package sse

import (
	"encoding/json"
	"net/http"
	"strconv"
	"sync"
	"time"
)

type client struct {
	id string
	w  http.ResponseWriter
	fl http.Flusher
}

var (
	roomsMu sync.Mutex
	rooms   = map[string]map[*client]struct{}{}
)

// Register 注册 SSE 相关路由：订阅、查询房间、发送信令。
func Register(mux *http.ServeMux) {
	mux.HandleFunc("/sse.signaling", subscribe)
	mux.HandleFunc("/api.room", getRoom)
	mux.HandleFunc("/api.signaling", postSignal)
}

// subscribe 处理 SSE 订阅，维持心跳并广播加入/离开事件。
func subscribe(w http.ResponseWriter, r *http.Request) {
	roomId := r.URL.Query().Get("roomId")
	clientId := r.URL.Query().Get("clientId")
	if roomId == "" || clientId == "" {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]any{"ok": false, "message": "roomId 和 clientId 必填"})
		return
	}
	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")
	w.Header().Set("Access-Control-Allow-Origin", "*")
	fl, _ := w.(http.Flusher)
	fl.Flush()

	c := &client{id: clientId, w: w, fl: fl}
	addClient(roomId, c)
	sendJSON(c, map[string]any{"type": "hello", "roomId": roomId, "clientId": clientId, "ts": time.Now().UnixMilli()})
	broadcast(roomId, map[string]any{"type": "join", "clientId": clientId, "ts": time.Now().UnixMilli()}, clientId, "")

	ticker := time.NewTicker(15 * time.Second)
	defer ticker.Stop()
	done := r.Context().Done()
	for {
		select {
		case <-ticker.C:
			_, _ = w.Write([]byte(": ping " + strconv.FormatInt(time.Now().UnixMilli(), 10) + "\n\n"))
			fl.Flush()
		case <-done:
			removeClient(roomId, c)
			broadcast(roomId, map[string]any{"type": "leave", "clientId": clientId, "ts": time.Now().UnixMilli()}, clientId, "")
			return
		}
	}
}

// getRoom 返回房间内当前用户列表。
func getRoom(w http.ResponseWriter, r *http.Request) {
	roomId := r.URL.Query().Get("roomId")
	if roomId == "" {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]any{"ok": false, "message": "roomId 必填"})
		return
	}
	roomsMu.Lock()
	set := rooms[roomId]
	users := make([]string, 0)
	for c := range set {
		users = append(users, c.id)
	}
	roomsMu.Unlock()
	json.NewEncoder(w).Encode(map[string]any{"ok": true, "users": users})
}

// postSignal 在房间内转发信令消息，支持定向发送。
func postSignal(w http.ResponseWriter, r *http.Request) {
	var body struct {
		RoomId, From, To, Type string
		Payload                any
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]any{"ok": false, "message": "roomId, from, type 必填"})
		return
	}
	if body.RoomId == "" || body.From == "" || body.Type == "" {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]any{"ok": false, "message": "roomId, from, type 必填"})
		return
	}
	delivered := broadcast(body.RoomId, map[string]any{"from": body.From, "type": body.Type, "payload": body.Payload, "ts": time.Now().UnixMilli()}, body.From, body.To)
	json.NewEncoder(w).Encode(map[string]any{"ok": true, "delivered": delivered})
}

// addClient 将客户端加入房间集合。
func addClient(roomId string, c *client) {
	roomsMu.Lock()
	defer roomsMu.Unlock()
	set := rooms[roomId]
	if set == nil {
		set = map[*client]struct{}{}
		rooms[roomId] = set
	}
	set[c] = struct{}{}
}

// removeClient 将客户端从房间集合移除，空房间自动清理。
func removeClient(roomId string, c *client) {
	roomsMu.Lock()
	defer roomsMu.Unlock()
	set := rooms[roomId]
	if set == nil {
		return
	}
	delete(set, c)
	if len(set) == 0 {
		delete(rooms, roomId)
	}
}

// sendJSON 以 SSE 数据格式发送 JSON。
func sendJSON(c *client, v any) {
	b, _ := json.Marshal(v)
	c.w.Write([]byte("data: "))
	c.w.Write(b)
	c.w.Write([]byte("\n\n"))
	c.fl.Flush()
}

// broadcast 向房间内所有或指定用户广播消息，返回送达数。
func broadcast(roomId string, msg any, excludeId, toId string) int {
	roomsMu.Lock()
	set := rooms[roomId]
	roomsMu.Unlock()
	if set == nil {
		return 0
	}
	delivered := 0
	for c := range set {
		if excludeId != "" && c.id == excludeId {
			continue
		}
		if toId != "" && c.id != toId {
			continue
		}
		sendJSON(c, msg)
		delivered++
	}
	return delivered
}
